x/liquidstake/keeper/liquidstake.go

```diff
@@ -0,0 +1,646 @@
+// It is used only for calculation and query and is not stored in kv.
+func (k Keeper) GetNetAmountState(ctx sdk.Context) (nas types.NetAmountState) {
+	totalRemainingRewards, totalDelShares, totalLiquidTokens := k.CheckDelegationStates(ctx, types.LiquidStakeProxyAcc)
+
+	totalUnbondingBalance := sdk.ZeroInt()
+	ubds := k.stakingKeeper.GetAllUnbondingDelegations(ctx, types.LiquidStakeProxyAcc)
+	for _, ubd := range ubds {
+		for _, entry := range ubd.Entries {
+			// use Balance(slashing applied) not InitialBalance(without slashing)
+			totalUnbondingBalance = totalUnbondingBalance.Add(entry.Balance)
+		}
+	}
+
+	nas = types.NetAmountState{
+		StkxprtTotalSupply:    k.bankKeeper.GetSupply(ctx, k.LiquidBondDenom(ctx)).Amount,
+		TotalDelShares:        totalDelShares,
+		TotalLiquidTokens:     totalLiquidTokens,
+		TotalRemainingRewards: totalRemainingRewards,
+		TotalUnbondingBalance: totalUnbondingBalance,
+		ProxyAccBalance:       k.GetProxyAccBalance(ctx, types.LiquidStakeProxyAcc).Amount,
+	}
+
+	nas.NetAmount = nas.CalcNetAmount()
+	nas.MintRate = nas.CalcMintRate()
+	return
+}
+
+// LiquidStake mints stkXPRT worth of staking coin value according to NetAmount and performs LiquidDelegate.
+func (k Keeper) LiquidStake(
+	ctx sdk.Context, proxyAcc, liquidStaker sdk.AccAddress, stakingCoin sdk.Coin) (newShares sdk.Dec, stkXPRTMintAmount math.Int, err error) {
+	params := k.GetParams(ctx)
+
+	// check minimum liquid staking amount
+	if stakingCoin.Amount.LT(params.MinLiquidStakeAmount) {
+		return sdk.ZeroDec(), sdk.ZeroInt(), types.ErrLessThanMinLiquidStakeAmount
+	}
+
+	// check bond denomination
+	bondDenom := k.stakingKeeper.BondDenom(ctx)
+	if stakingCoin.Denom != bondDenom {
+		return sdk.ZeroDec(), sdk.ZeroInt(), sdkerrors.Wrapf(
+			types.ErrInvalidBondDenom, "invalid coin denomination: got %s, expected %s", stakingCoin.Denom, bondDenom,
+		)
+	}
+
+	whitelistedValsMap := types.GetWhitelistedValsMap(params.WhitelistedValidators)
+	activeVals := k.GetActiveLiquidValidators(ctx, whitelistedValsMap)
+	if activeVals.Len() == 0 || !activeVals.TotalWeight(whitelistedValsMap).IsPositive() {
+		return sdk.ZeroDec(), sdk.ZeroInt(), types.ErrActiveLiquidValidatorsNotExists
+	}
+
+	// NetAmount must be calculated before send
+	nas := k.GetNetAmountState(ctx)
+
+	// send staking coin to liquid staking proxy account to proxy delegation, need sufficient spendable balances
+	err = k.bankKeeper.SendCoins(ctx, liquidStaker, proxyAcc, sdk.NewCoins(stakingCoin))
+	if err != nil {
+		return sdk.ZeroDec(), sdk.ZeroInt(), err
+	}
+
+	// mint stkxprt, MintAmount = TotalSupply * StakeAmount/NetAmount
+	liquidBondDenom := k.LiquidBondDenom(ctx)
+	stkXPRTMintAmount = stakingCoin.Amount
+	if nas.StkxprtTotalSupply.IsPositive() {
+		stkXPRTMintAmount = types.NativeTokenToStkXPRT(stakingCoin.Amount, nas.StkxprtTotalSupply, nas.NetAmount)
```

Comment:
Why not stakingCoin.Amount * nas.MintRate ?

Reply:
It's the same formula, but `NativeTokenToStkXPRT` wraps some tricky truncation stuff.

`MintRate` already truncated after `/ netAmount`, while in `NativeTokenToStkXPRT` above the truncation happens only after result of `stakingCoin.Amount * totalSupply` which gives it better precision.

Not sure if the precision of these calculations could be improved further.